Source/Tools/FEXServer/Unwind/ELFMapping.cpp

```diff
@@ -0,0 +1,429 @@
+#include <FEXCore/Utils/LogManager.h>
+
+#include "ELFMapping.h"
+#include "FileMapping.h"
+#include <cassert>
+#include <cstring>
+#include <elf.h>
+#include <map>
+#include <sys/mman.h>
+#include <sys/stat.h>
+
+namespace ELFMapping {
+  struct ELFMemMapping {
+    int FD;
+    uint64_t OriginalBase;
+    char *BaseRemapped;
+    size_t TotalSize;
+
+    struct MappedRemapping {
+      uint64_t OriginalBase;
+      void* Remapped;
+      size_t Size;
+    };
+    std::vector<MappedRemapping> ReadMappings;
+
+    union {
+      Elf32_Ehdr _32;
+      Elf64_Ehdr _64;
+    } Header;
+
+    union SectionHeader {
+      const Elf32_Shdr *_32;
+      const Elf64_Shdr *_64;
+    };
+
+    union ProgramHeader {
+      const Elf32_Phdr *_32;
+      const Elf64_Phdr *_64;
+    };
+
+    std::vector<SectionHeader> SectionHeaders;
+    std::vector<ProgramHeader> ProgramHeaders;
+    std::vector<ELFSymbol> Symbols;
+    std::vector<std::string> AddressSymbolNames;
+    std::vector<uintptr_t> UnwindEntries;
+    std::unordered_map<std::string, ELFSymbol *> SymbolMap;
+    std::map<uint64_t, ELFSymbol *> SymbolMapByAddress;
+  };
+
+  uint64_t IsELFBits(ELFMemMapping *Mapping) {
+    constexpr size_t ELFHeaderSize = std::max(sizeof(Elf32_Ehdr), sizeof(Elf64_Ehdr));
+    if (Mapping->TotalSize < ELFHeaderSize) {
+      return 0;
+    }
+
+    const uint8_t* Data = reinterpret_cast<const uint8_t*>(Mapping->BaseRemapped);
+
+    if (Data[EI_MAG0] != ELFMAG0 ||
+        Data[EI_MAG1] != ELFMAG1 ||
+        Data[EI_MAG2] != ELFMAG2 ||
+        Data[EI_MAG3] != ELFMAG3) {
+      return 0;
+    }
+
+    if (Data[EI_CLASS] == ELFCLASS32) {
+      const Elf32_Ehdr *Header = reinterpret_cast<const Elf32_Ehdr *>(Data);
+      if (Header->e_machine == EM_386) {
+        return 32;
+      }
+    }
+    else if (Data[EI_CLASS] == ELFCLASS64) {
+      const Elf64_Ehdr *Header = reinterpret_cast<const Elf64_Ehdr *>(Data);
+      if (Header->e_machine == EM_X86_64) {
+        return 64;
+      }
+    }
+
+    return 0;
+  }
+
+  void ParseELFSymbols64(ELFMemMapping *Mapping) {
+    const char* RawBase = reinterpret_cast<const char*>(Mapping->BaseRemapped);
+    memcpy(&Mapping->Header, RawBase, sizeof(Elf64_Ehdr));
+    Mapping->SectionHeaders.resize(Mapping->Header._64.e_shnum);
+    Mapping->ProgramHeaders.resize(Mapping->Header._64.e_phnum);
+
+    const Elf64_Shdr *RawShdrs =
+      reinterpret_cast<const Elf64_Shdr *>(&RawBase[Mapping->Header._64.e_shoff]);
+    const Elf64_Phdr *RawPhdrs =
+      reinterpret_cast<const Elf64_Phdr *>(&RawBase[Mapping->Header._64.e_phoff]);
+
+    if (Mapping->Header._64.e_shoff > Mapping->TotalSize) {
+      __builtin_unreachable();
```

Comment:
Looks like this could be reached if the input is erroneous/corrupted. `assert` or another panic-ing function would be more fitting.

Reply:
Returned early safely so we don't bring down the FEXServer on ELF parsing.